underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AUnderFileSystem.java

```diff
@@ -0,0 +1,725 @@
+      path = stripPrefixIfPresent(path);
+      return new S3AInputStream(mBucketName, path, mClient);
+    } catch (AmazonClientException e) {
+      LOG.error("Failed to open file: {}", path, e);
+      return null;
+    }
+  }
+
+  /**
+   * Opens a S3 object at given position and returns the opened input stream.
+   *
+   * @param path the S3 object path
+   * @param pos the position to open at
+   * @return the opened input stream
+   * @throws java.io.IOException if failed to open file at position
+   */
+  public InputStream openAtPosition(String path, long pos) throws IOException {
+    try {
+      path = stripPrefixIfPresent(path);
+      return new S3AInputStream(mBucketName, path, mClient, pos);
+    } catch (AmazonClientException e) {
+      LOG.error("Failed to open file {} at position {}:", path, pos, e);
+      return null;
+    }
+  }
+
+  @Override
+  public boolean rename(String src, String dst) throws IOException {
+    if (!exists(src)) {
+      LOG.error("Unable to rename {} to {} because source does not exist.", src, dst);
+      return false;
+    }
+    if (exists(dst)) {
+      LOG.error("Unable to rename {} to {} because destination already exists.", src, dst);
+      return false;
+    }
+    // Source exists and destination does not exist
+    if (isFolder(src)) {
+      // Rename the source folder first
+      if (!copy(convertToFolderName(src), convertToFolderName(dst))) {
+        return false;
+      }
+      // Rename each child in the src folder to destination/child
+      String [] children = list(src);
+      for (String child: children) {
+        if (!rename(PathUtils.concatPath(src, child), PathUtils.concatPath(dst, child))) {
+          return false;
+        }
+      }
+      // Delete src and everything under src
+      return delete(src, true);
+    }
+    // Source is a file and Destination does not exist
+    return copy(src, dst) && deleteInternal(src);
+  }
+
+  // Not supported
+  @Override
+  public void setConf(Object conf) {}
+
+  // Not supported
+  @Override
+  public void setOwner(String path, String user, String group) {}
+
+  // Not supported
+  @Override
+  public void setMode(String path, short mode) throws IOException {}
+
+  // Not supported
+  @Override
+  public String getOwner(String path) throws IOException {
+    return "";
+  }
+
+  // Not supported
+  @Override
+  public String getGroup(String path) throws IOException {
+    return "";
+  }
+
+  // Not supported
+  @Override
+  public short getMode(String path) throws IOException {
+    return Constants.DEFAULT_FILE_SYSTEM_MODE;
+  }
+
+  /**
+   * Appends the directory suffix to the key.
+   *
+   * @param key the key to convert
+   * @return key as a directory path
+   */
+  private String convertToFolderName(String key) {
+    // Strips the slash if it is the end of the key string. This is because the slash at
+    // the end of the string is not part of the Object key in S3.
+    if (key.endsWith(PATH_SEPARATOR)) {
```

Comment:
@madanadit has added `CommonUtils.stripeSuffixIfPresent` which you may want to use when his PR gets merged